### PR TITLE
Introduce support for plugins

### PR DESCRIPTION
**Pull request by marcospassos**
## Summary
Introduce support for plugins.

### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: src/globalPlug.ts [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+            if (typeof controller.enable === 'function') {
+                const promise = controller.enable();
+
+                if (promise instanceof Promise) {
+                    initialization.push(
+                        promise
+                            .then(() => logger.debug(`Plugin "${pluginName}" enabled`))
+                            .catch(() => logger.error(`Failed to enable plugin "${pluginName}"`)),
+                    );
+                }
```
Comment (Fryuni) [inline src/globalPlug.ts:80]: What happens if there is an enable function but it doesn't return a promise?
The signature is `() => Promise<void>|void`, it should be logged either way.

File: src/globalPlug.ts [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+            const promise = controller.disable();
+
+            if (promise instanceof Promise) {
+                pending.push(
+                    promise
+                        .then(() => logger.debug(`Plugin "${pluginName}" disabled`))
+                        .catch(() => logger.error(`Failed to disable "${pluginName}"`)),
+                );
+            }
```
Comment (Fryuni) [inline src/globalPlug.ts:174]: Also here it can should be logged even if its not a Promise

File: src/globalPlug.ts (hunk 1 above)
Comment (marcospassos) [reply src/globalPlug.ts:80]: Good catch

File: src/globalPlug.ts (hunk 2 above)
Comment (marcospassos) [reply src/globalPlug.ts:174]: Agreed

**Pull request merged**